From 00333182aa303b29975ac3aa656eaf29c4d5d0e5 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?=D0=A0=D1=83=D1=81=D0=BB=D0=B0=D0=BD=20=D0=98=D0=B6=D0=B1?=
 =?UTF-8?q?=D1=83=D0=BB=D0=B0=D1=82=D0=BE=D0=B2?= <lrn1986@gmail.com>
Date: Tue, 14 Apr 2015 14:51:22 +0000
Subject: [PATCH 19/20] [W32] Update windowsVersion() and
 osVersionForUAString()

Add support for newer Windows versions to windowsVersion(), handle all
defined cases in osVersionForUAString(), avoiding warnings like
../webkitgtk-2.4.8/Source/WebCore/platform/win/SystemInfo.cpp: In function 'WTF::String WebCore::osVersionForUAString()':
../webkitgtk-2.4.8/Source/WebCore/platform/win/SystemInfo.cpp:94:12: warning: enumeration value 'WindowsCE5' not handled in switch [-Wswitch]
repeated for each unhandled value
---
 Source/WebCore/platform/win/SystemInfo.cpp | 107 +++++++++++++++++++++++++++--
 Source/WebCore/platform/win/SystemInfo.h   |   9 +++
 2 files changed, 111 insertions(+), 5 deletions(-)

diff --git a/Source/WebCore/platform/win/SystemInfo.cpp b/Source/WebCore/platform/win/SystemInfo.cpp
index c9edca0..de5e8dc 100644
--- a/Source/WebCore/platform/win/SystemInfo.cpp
+++ b/Source/WebCore/platform/win/SystemInfo.cpp
@@ -29,6 +29,10 @@
 #include <windows.h>
 #include <wtf/text/WTFString.h>
 
+#ifndef SM_SERVERR2
+#define SM_SERVERR2 89
+#endif
+
 namespace WebCore {
 
 WindowsVersion windowsVersion(int* major, int* minor)
@@ -65,15 +69,73 @@ WindowsVersion windowsVersion(int* major, int* minor)
                 version = (minorVersion == 10) ? Windows98 : WindowsME;
         } else {
             if (majorVersion == 5) {
-                if (!minorVersion)
+                switch (minorVersion) {
+                default:
+                case 0:
                     version = Windows2000;
+                    break;
+                case 1:
+                    version = WindowsXP;
+                    break;
+                case 2:
+                    if (versionInfo.wProductType == VER_NT_WORKSTATION)
+                        version = WindowsXP64;
+                    else {
+                        if (GetSystemMetrics(SM_SERVERR2) == 0)
+                            version = WindowsServer2003;
+                        else
+                            version = WindowsServer2003R2;
+                    }
+                    break;
+                }
+            } else if (majorVersion == 6) {
+                if (versionInfo.wProductType == VER_NT_WORKSTATION)
+                    switch (minorVersion) {
+                    default:
+                    case 0:
+                        version = WindowsVista;
+                        break;
+                    case 1:
+                        version = Windows7;
+                        break;
+                    case 2:
+                        version = Windows8;
+                        break;
+                    case 3:
+                        version = Windows81;
+                        break;
+                    }
                 else
-                    version = (minorVersion == 1) ? WindowsXP : WindowsServer2003;
-            } else if (majorVersion >= 6) {
+                    switch (minorVersion) {
+                    default:
+                    case 0:
+                        version = WindowsServer2008;
+                        break;
+                    case 1:
+                        version = WindowsServer2008R2;
+                        break;
+                    case 2:
+                        version = WindowsServer2012;
+                        break;
+                    case 3:
+                        version = WindowsServer2012R2;
+                        break;
+                    }
+            } else if (majorVersion == 10) {
                 if (versionInfo.wProductType == VER_NT_WORKSTATION)
-                    version = (majorVersion == 6 && !minorVersion) ? WindowsVista : Windows7;
+                    switch (minorVersion) {
+                    default:
+                    case 0:
+                        version = Windows10;
+                        break;
+                    }
                 else
-                    version = WindowsServer2008;
+                    switch (minorVersion) {
+                    default:
+                    case 0:
+                        version = WindowsServer2016;
+                        break;
+                    }
             } else
                 version = (majorVersion == 4) ? WindowsNT4 : WindowsNT3;
         }
@@ -97,6 +159,9 @@ static String osVersionForUAString()
     case WindowsCE3:
         return "Windows CE";
     case WindowsCE4:
+    case WindowsCE5:
+    case WindowsCE6:
+    case WindowsCE7:
         return "Windows CE .NET";
     case Windows3_1:
         return "Windows 3.1";
@@ -106,8 +171,40 @@ static String osVersionForUAString()
         return "Windows 98";
     case WindowsME:
         return "Windows 98; Win 9x 4.90";
+    case WindowsNT3:
+        return "WinNT3.0";
     case WindowsNT4:
         return "WinNT4.0";
+    case Windows2000:
+        return "Windows 2000";
+    case WindowsXP:
+        return "Windows XP";
+    case WindowsServer2003:
+        return "Windows Server 2003";
+    case WindowsXP64:
+        return "Windows XP Professional x64 Edition";
+    case WindowsServer2003R2:
+        return "Windows Server 2003 R2";
+    case WindowsVista:
+        return "Windows Vista";
+    case WindowsServer2008:
+        return "Windows Server 2008";
+    case Windows7:
+        return "Windows 7";
+    case WindowsServer2008R2:
+        return "Windows Server 2008 R2";
+    case Windows8:
+        return "Windows 8";
+    case WindowsServer2012:
+        return "Windows Server 2012";
+    case Windows81:
+        return "Windows 8.1";
+    case WindowsServer2012R2:
+        return "Windows Server 2012 R2";
+    case Windows10:
+        return "Windows 10";
+    case WindowsServer2016:
+        return "Windows Server 2016";
     }
 
     const char* familyName = (version >= WindowsNT3) ? "Windows NT " : "Windows CE ";
diff --git a/Source/WebCore/platform/win/SystemInfo.h b/Source/WebCore/platform/win/SystemInfo.h
index 2631ace..7bfb222 100644
--- a/Source/WebCore/platform/win/SystemInfo.h
+++ b/Source/WebCore/platform/win/SystemInfo.h
@@ -54,9 +54,18 @@ enum WindowsVersion {
     Windows2000,
     WindowsXP,
     WindowsServer2003,
+    WindowsXP64,
+    WindowsServer2003R2,
     WindowsVista,
     WindowsServer2008,
     Windows7,
+    WindowsServer2008R2,
+    Windows8,
+    WindowsServer2012,
+    Windows81,
+    WindowsServer2012R2,
+    Windows10,
+    WindowsServer2016,
 };
 
 // If supplied, |major| and |minor| are set to the OSVERSIONINFO::dwMajorVersion
-- 
1.8.5.3

